pkg/store/proxy_heap.go

```diff
@@ -96,11 +99,16 @@ func (d *dedupResponseHeap) Next() bool {
 		}
 
 		lbls := resp.GetSeries().Labels
-		lastLbls := d.responses[len(d.responses)-1].GetSeries().Labels
+		lastResponse := d.responses[len(d.responses)-1]
+		lastLbls := lastResponse.GetSeries().Labels
 
-		if labels.Compare(labelpb.ZLabelsToPromLabels(lbls), labelpb.ZLabelsToPromLabels(lastLbls)) == 0 {
+		// Unless a response comes from query pushdown, it is safe to merge
+		// chunks from replica series together into one slice.
+		labelsEqual := labels.Compare(labelpb.ZLabelsToPromLabels(lbls), labelpb.ZLabelsToPromLabels(lastLbls)) == 0
```

Comment:
I would love to use `labelsEqual = resp.signature == lastResponse.signature` here, but unfortunately it breaks tests with overlapping and interleaved chunks from prometheus replicas. This was introduced and solved in:
* https://github.com/thanos-io/thanos/issues/2401
* https://github.com/thanos-io/thanos/pull/2522

Reply:
I tested that change in a canary environment, and together with the new engine saw a decrease in a query from 75s to 8s, which is just insane. We have a layered querier setup, so deduplicating as soon as possible makes a massive difference.

Reply:
<img src="https://media1.giphy.com/media/8b9Xax6L7qtAkAimGm/giphy.gif"/>